gc.c

```diff
@@ -3220,28 +3220,84 @@ struct each_obj_args {
     void *data;
 };
 
+struct each_obj_data {
+    struct each_obj_args args;
+    struct heap_page **pages;
+    size_t num_pages;
+};
+
+static bool
+is_ptr_to_heap_page(rb_objspace_t *objspace, struct heap_page *page)
+{
+    for (size_t i = 0; i < heap_allocated_pages; i++) {
+        if (heap_pages_sorted[i] == page) {
+            return true;
+        }
+    }
```

Comment:
Can we make this more efficient?  We know the `heap_pages_sorted` list is sorted by address, so I think we could probably use a binary search here.  What do you think?

Reply:
`heap_pages_sorted` is a list of `struct heap_page` pointers sorted by the start address of the page. But the `page` pointer here we can't sure if it's safe to read from (since it could be free'd memory) so we don't know the start address of this page so we can't do binary search in `heap_pages_sorted`.

Alternatively, we could also store both the pointer to the `struct heap_page` and the pointer to the start of the page in the malloc'd buffer and that way we can do binary search here. WDYT?

Reply:
Alternatively, what we could also do, is this:

1. Copy `heap_pages_sorted`.
2. Traverse the linked list of pages on the eden heap, skipping ones that don't exist in the buffer we created in step 1.

This way we can still perform binary search in step 2 since the buffer in step 1 is sorted.

Reply:
I like the idea of copying `heap_pages_sorted`. It makes sense!